learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,287 @@
+        return f
+
+    return decorator
+
+def make_ajax_raw_handler(raw_ajax_function, remote_url):
+    async def ajax_passthrough(request):
+        runtime = learning_observer.runtime.Runtime(request)
+        response = await raw_ajax_function(runtime, remote_url, retry=True, **request.match_info)
+        return aiohttp.web.json_response(response)
+    return ajax_passthrough
+
+def make_cleaner_handler(raw_function, cleaner_function, name=None):
+    async def cleaner_handler(request):
+        response = cleaner_function(await raw_function(request, **request.match_info))
+        if isinstance(response, dict) or isinstance(response, list):
+            return aiohttp.web.json_response(response)
+        elif isinstance(response, str):
+            return aiohttp.web.Response(text=response)
+        else:
+            raise AttributeError(f"Invalid response type: {type(response)}")
+    if name is not None:
+        setattr(cleaner_handler, "__qualname__", name + "_handler")
+
+    return cleaner_handler
+
+def make_cleaner_function(raw_function, cleaner_function, name=None):
+    async def cleaner_local(request, **kwargs):
+        lms_response = await raw_function(request, **kwargs)
+        clean = cleaner_function(lms_response)
+        return clean
+    if name is not None:
+        setattr(cleaner_local, "__qualname__", name)
+    return cleaner_local
+
+async def raw_ajax(runtime, target_url, lms_name, base_url=None, **kwargs):
+    """
+    Make an authenticated AJAX call to a specified service (e.g., Google, Canvas), handling 
+    authorization, caching, and retries.
+
+    Parameters:
+    - runtime: An instance of the Runtime class containing request information.
+    - lms_name: A string indicating the name of the service ('google' or 'canvas').
+    - target_url: The URL endpoint to be called, with optional formatting using kwargs.
+    - base_url: An optional base URL for the service. If provided, it will be prefixed 
+      to target_url.
+    - kwargs: Additional keyword arguments to format the target_url or control behavior 
+      (e.g., retry).
+
+    Returns:
+    - A JSON response from the requested service, with key translation if necessary.
+
+    Raises:
+    - aiohttp.web.HTTPUnauthorized: If the request lacks necessary authorization.
+    - aiohttp.ClientResponseError: If the request fails, with special handling for 401 errors on Canvas.
+    """
+    # Retrieve the incoming request and active user
+    request = runtime.get_request()
+    user = await learning_observer.auth.get_active_user(request)
+    
+    # Extract 'retry' flag from kwargs (defaults to False)
+    retry = kwargs.pop('retry', False)
+    
+    # Retrieve session and determine the appropriate headers based on the service
+    session = await aiohttp_session.get_session(request)
+    headers = {
+        'google': request[constants.AUTH_HEADERS],
+        'canvas': session.get(constants.CANVAS_AUTH_HEADERS)
```

Comment:
Not a fan of these being stored in different places. Why is one session and the other request?

Reply:
Fixed it